wgsl/index.bs

```diff
@@ -581,28 +615,300 @@ storage_class
+of host-shareable type |S| in storage class |C|,
+where |C| is either `uniform` or `storage`.
+It is defined recursively in the following table:
 
-TODO: *This is a stub*: Collectively, the <dfn noexport>stride</dfn> and <dfn noexport>offset</dfn> attributes are called *layout attributes*.
+<table class='data'>
+  <caption>Allocation extent of a value of host-shareable type</caption>
+  <thead>
+    <tr><th>Host-shareable type |S|
+        <th>|Extent|(|V|,`storage`)<br>where |V| is of type |S|
+        <th>|Extent|(|V|,`uniform`)<br>where |V| is of type |S|
+  </thead>
+  <tr><td>[=i32=], [=u32=], or [=f32=]
+      <td>4
+      <td>4
+  <tr><td>vec|N|&lt;|T|&gt;,
+      where |T| is one of [=i32=], [=u32=], or [=f32=]
+      <td>|N| &times; 4
+      <td>|N| &times; 4
+  <tr algorithm="extent of a matrix with 2 rows">
+      <td>mat|N|x2&lt;f32&gt;
+      <td>|N| &times; 8
+      <td>|N| &times; 16
+  <tr algorithm="extent of a matrix with 3 rows">
+      <td>mat|N|x3&lt;f32&gt;
+      <td>|N| &times; 16
+      <td>|N| &times; 16
+  <tr algorithm="extent of a matrix with 4 rows">
+      <td>mat|N|x4&lt;f32&gt;
+      <td>|N| &times; 16
+      <td>|N| &times; 16
+  <tr algorithm="extent of an array">
+      <td>array&lt;<var ignore>T</var>,|N|>
+      <td>|N| &times; |Stride|(|S|)
+      <td>|N| &times; |Stride|(|S|)
+  <tr algorithm="extent of an runtime-sized array">
+      <td>array&lt;<var ignore>T</var>&gt;
+      <td>|N|<sub>runtime</sub> &times; |Stride|(|S|),<br>
+where |N|<sub>runtime</sub> is the runtime-determined number of elements of <var ignore>V</var>
+      <td>Not applicable:  [=runtime-sized=] arrays cannot appear in `uniform` storage
+  <tr algorithm="extent of a structure">
+      <td>struct&lt;|T|<sub>1</sub>,...,|T|<sub>|n|</sub>&gt;
+      <td>[=roundUp=](|Align|(|S|,`storage`),|L|),<br>
+      where |L|&nbsp;=&nbsp;|Offset|(|S|,|n|)&nbsp;+&nbsp;|Extent|(|V|<sub>|n|</sub>,`storage`)),<br>
+      and |V|<sub>|n|</sub> is the last member of |V|
+      <td>[=roundUp=](|Align|(|S|,`uniform`),|L|),<br>
+      where |L|&nbsp;=&nbsp;|Offset|(|S|,|n|)&nbsp;+&nbsp;|Extent|(|V|<sub>|n|</sub>,`uniform`)),<br>
+      and |V|<sub>|n|</sub> is the last member of |V|
+</table>
 
-TODO: <dfn noexport>layout attribute</dfn>
+When a type |S| is not a runtime-sized array and it does not contain a runtime-sized array, then
+all values |V| of type |S| will have the same allocation extent for a storage class |C|.
+In these cases we define the allocation extent of the *type* |S| as that common value:
+|Extent|(|S|,|C|) = |Extent|(|V|,|C|), for any |V| of type |S|.
+
+Note: When underlying the target is a Vulkan device, we assume the device does
+not support the `scalarBlockLayout` feature.
+Therefore, a data value must not be placed in the padding at the end of a structure or matrix,
+nor in the padding at the last element of an array.
+Counting the such padding as part of the allocation extent allows [SHORTNAME] to capture this constraint.
```

Comment:
nit: `Counting the padding` or `Counting such padding`

Reply:
Fixed in 9603cf